lib/charms/postgresql_k8s/v0/postgresql.py

```diff
@@ -0,0 +1,170 @@
+        self.password = password
+        self.database = database
+
+    def _connect_to_database(self, database: str = None) -> psycopg2.extensions.connection:
+        """Creates a connection to the database.
+
+        Args:
+            database: database to connect to (defaults to the database
+                provided when the object for this class was created).
+
+        Returns:
+             psycopg2 connection object.
+        """
+        connection = psycopg2.connect(
+            f"dbname='{database if database else self.database}' user='{self.user}' host='{self.host}' password='{self.password}' connect_timeout=1"
+        )
+        connection.autocommit = True
+        return connection
+
+    def create_database(self, database: str, user: str) -> None:
+        """Creates a new database and grant privileges to a user on it.
+
+        Args:
+            database: database to be created.
+            user: user that will have access to the database.
+        """
+        try:
+            connection = self._connect_to_database()
+            cursor = connection.cursor()
+            cursor.execute(f"SELECT datname FROM pg_database WHERE datname='{database}';")
+            if cursor.fetchone() is None:
+                cursor.execute(sql.SQL("CREATE DATABASE {};").format(sql.Identifier(database)))
+            cursor.execute(
+                sql.SQL("GRANT ALL PRIVILEGES ON DATABASE {} TO {};").format(
+                    sql.Identifier(database), sql.Identifier(user)
+                )
+            )
+        except psycopg2.Error as e:
+            logger.error(f"Failed to create database: {e}")
+            raise PostgreSQLCreateDatabaseError()
+
+    def create_user(self, user: str, password: str, admin: bool = False) -> None:
+        """Creates a database user.
+
+        Args:
+            user: user to be created.
+            password: password to be assigned to the user.
+            admin: whether the user should have additional admin privileges.
+        """
+        try:
+            with self._connect_to_database() as connection, connection.cursor() as cursor:
+                cursor.execute(f"SELECT TRUE FROM pg_roles WHERE rolname='{user}';")
+                user_definition = f"{user} WITH LOGIN{' SUPERUSER' if admin else ''} ENCRYPTED PASSWORD '{password}'"
+                if cursor.fetchone() is not None:
+                    cursor.execute(f"ALTER ROLE {user_definition};")
+                else:
+                    cursor.execute(f"CREATE ROLE {user_definition};")
+        except psycopg2.Error as e:
+            logger.error(f"Failed to create user: {e}")
+            raise PostgreSQLCreateUserError()
+
+    def delete_user(self, user: str) -> None:
+        """Deletes a database user.
+
+        Args:
+            user: user to be deleted.
+        """
+        # List all databases.
+        try:
+            with self._connect_to_database() as connection, connection.cursor() as cursor:
```

Comment:
is there a reason why there are 3 calls to `self._connect_to_database` in this method? is to break up the method logic into 3 transactions?